asmcomp/arm/emit.mlp

```diff
@@ -796,30 +801,48 @@ let emit_instr i =
           `	pop	\{trap_ptr, pc}\n`; 2
         end
 
+(* Upper bound on the size of the code sequence for a Linear instruction,
+   in 32-bit words. *)
+
+let max_instruction_size i =
+  match i.desc with
+  | Lswitch jumptbl ->
+      if !arch > ARMv6 && !thumb
+      then 1 + (Array.length jumptbl + 1) / 2 + Array.length jumptbl
+      else 2 + Array.length jumptbl
+  | _ ->
+      4   (* conservative upper bound *)
+
 (* Emission of an instruction sequence *)
 
-let rec emit_all ninstr i =
+let rec emit_all ninstr fallthrough i =
+  (* ninstr = number of 32-bit code words emitted since last constant island *)
+  (* fallthrough is true if previous instruction can fall through *)
   if i.desc = Lend then () else begin
-    let n = emit_instr i in
-    let ninstr' = ninstr + n in
+    (* Make sure literals not yet emitted remain addressable,
+       or emit them in a new constant island. *)
     (* fldd can address up to +/-1KB, ldr can address up to +/-4KB *)
     let limit = (if !fpu >= VFPv2 && !float_literals <> []
                  then 127
                  else 511) in
-    let limit = limit - !num_literals in
-    if ninstr' >= limit - 64 && not(has_fallthrough i.desc) then begin
-      emit_literals();
-      emit_all 0 i.next
-    end else if !num_literals != 0 && ninstr' >= limit then begin
-      let lbl = new_label() in
-      `	b	{emit_label lbl}\n`;
-      emit_literals();
-      `{emit_label lbl}:\n`;
-      emit_all 0 i.next
-    end else
-      emit_all ninstr' i.next
+    let limit = limit - !num_literals - max_instruction_size i in
```

Comment:
The name `num_literals` is confusing since it sounds like, well, the number of literals---but actually it's the number of words they occupy.  Could we rename this?